// @zerostep/playwright/lib/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOGS_ENABLED = exports.WEBSOCKET_URL = exports.WEBDRIVER_ELEMENT_KEY = exports.MAX_TASK_CHARS = exports.MAX_TASK_CHARS_STRING = exports.LOGS_ENABLED_STRING = exports.PACKAGE_NAME = exports.WEBSOCKET_HOST = exports.WEBSOCKET_PROTOCOL = exports.TOKEN = exports.getConfigValueString = exports.getConfigFile = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
let configFile = null;
const getConfigFile = () => {
    if (!configFile) {
        let cf = {};
        const configPath = path.join(process.cwd(), 'zerostep.config.json');
        if (fs.existsSync(configPath)) {
            cf = JSON.parse(fs.readFileSync(configPath, 'utf8'));
        }
        configFile = cf;
    }
    return configFile;
};
exports.getConfigFile = getConfigFile;
const getConfigValueString = (key, fallback) => {
    if (process.env[`ZEROSTEP_${key}`] !== undefined) {
        return process.env[`ZEROSTEP_${key}`];
    }
    const configFile = (0, exports.getConfigFile)();
    if (configFile[key] !== undefined) {
        return configFile[key].toString();
    }
    else if (fallback) {
        return fallback;
    }
    else {
        return '';
    }
};
exports.getConfigValueString = getConfigValueString;
// Configurable via environment variables or the config file
exports.TOKEN = (0, exports.getConfigValueString)('TOKEN');
exports.WEBSOCKET_PROTOCOL = (0, exports.getConfigValueString)('WEBSOCKET_PROTOCOL', 'https');
exports.WEBSOCKET_HOST = (0, exports.getConfigValueString)('WEBSOCKET_HOST', 'devai.us-east-1.reflect.run');
exports.PACKAGE_NAME = (0, exports.getConfigValueString)('PACKAGE_NAME', 'zerostep');
exports.LOGS_ENABLED_STRING = (0, exports.getConfigValueString)('LOGS_ENABLED', 'false');
exports.MAX_TASK_CHARS_STRING = (0, exports.getConfigValueString)('MAX_TASK_CHARS', '2000');
// Computed configuration values
exports.MAX_TASK_CHARS = parseInt(exports.MAX_TASK_CHARS_STRING);
exports.WEBDRIVER_ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf';
exports.WEBSOCKET_URL = `${exports.WEBSOCKET_PROTOCOL}://${exports.WEBSOCKET_HOST}/api?key=${exports.TOKEN}`;
exports.LOGS_ENABLED = exports.LOGS_ENABLED_STRING === 'true';
